[Review] feat (pubsub): SKS SecurityGroups

Muddasir-Shakil opened the pull request:
This PR is the second in series of SKS support. It adds the support for managing SecurityGroups on SKS server side. A securitygroup is an abstraction that represents the message security settings and security keys for a subset of NetworkMessages exchanged between Publishers and Subscribers. A SKS manages the SecurityGroups and maintains the mapping
between Roles and their access Permissions.

In the implementation two server API methods are added to create and remove the SecurityGroup in the server. With a securitygroup a keystorage with initial keys is also created in the server. The keystorage is updated periodically with new keys
by a callback function. The period of update cycle is set by KeyLifeTime parameter in SecurityGroup configuration. The conditional compilation with information model support is also added for SecurityGroup.

jpfr commented on src/pubsub/ua_pubsub_manager.c:
```diff
@@ -778,6 +778,14 @@ UA_PubSubManager_delete(UA_Server *server, UA_PubSubManager *pubSubManager) {
+    UA_SecurityGroup *tmpSG1, *tmpSG2;
+    TAILQ_FOREACH_SAFE(tmpSG1, &server->pubSubManager.securityGroups,
+                       listEntry, tmpSG2)
+        removeSecurityGroup(server, tmpSG1->securityGroupNodeId);
```
You already have the SG. No need to forward the id and do a second lookup.
removeSecurityGroup should take a pointer to the SG.
Any user of it can do a UA_SecurityGroup_findSGbyId before.
removeSecurityGroup then also does not have to return a statuscode. It isn't evaluated anyhow.

jpfr commented on src/pubsub/ua_pubsub_securitygroup.c:
```diff
@@ -0,0 +1,322 @@
+        TAILQ_INSERT_TAIL(&server->pubSubManager.securityGroups, newSecurityGroup,
+                          listEntry);
+    } else {
+        TAILQ_INIT(&server->pubSubManager.securityGroups);
```
You should call TAILQ_INIT in the _init of the PubSubManager.
Then you don't need to differentiate between cases here at all. Just do a INSERT_TAIL.

Muddasir-Shakil replied:
I anyhow have to check both cases because if list is empty then TAILQ requires INSERT_HEAD and in rest of the cases INSERT_TAIL is used.
Also I checkout other places where TAILQ_INIT is used such as shown in the following. TAILQ is used in this manner. So I simply used as it was used for other cases like connections
https://github.com/open62541/open62541/blob/779e6b7902ff223b27d402c952fc5f06e967dd23/src/pubsub/ua_pubsub_manager.c#L78

also if calling TAILQ_INIT in the _init of the PubSubManager, would result in initializing securitygroups list every time whether the Security Groups will be used or not. Therefore I did the initialization when the first SecurityGroup is added to the server only then the list should be initialized.

jpfr replied:
>  if list is empty then TAILQ requires INSERT_HEAD and in rest of the cases INSERT_TAIL is used.

What gives you that impression?
Reading the code in queue.h, INSERT_TAIL always works once the structure is correctly initialized.

Muddasir-Shakil replied:
> > if list is empty then TAILQ requires INSERT_HEAD and in rest of the cases INSERT_TAIL is used.
> 
> What gives you that impression? Reading the code in queue.h, INSERT_TAIL always works once the structure is correctly initialized.

I tested it and now it works with INSERT_TAIL. However, at other places as mentioned above it is used as the same way I did. May be it should be improved at other places like `UA_PubSubManager_addConnection`, and  `UA_PubSubManager_addTopic`

Regarding INIT, It added to pubsub UA_PubSubManager_init `UA_PubSubManager_init` with `#UA_ENABLE_SKS`. and I think this can done for other PubSub related lists.?

jpfr replied:
Ok. let's improve the other places then as well.
But not in this PR.

jpfr commented on src/pubsub/ua_pubsub_securitygroup.c:
```diff
@@ -0,0 +1,322 @@
+}
+
+static void
+updateSKSKeyStorage(UA_Server *server, UA_SecurityGroup *securityGroup){
```
Silent failures shouldn't happen.
If you cannot return a status code here, then use the logging from the server to warn users.
The goto error is only used once. That should be move inline to reduce the code flow complexity.

jpfr reviewed (commented):
Looks really good.
A few detail comments inline.
If you want us to preserve the commit history, then use the conventions from CONTRIBUTING.md.
Or see the examples in the master history.
Otherwise I would squash the PR into a single "feat(pubsub):" commit.

Muddasir-Shakil commented on src/pubsub/ua_pubsub_securitygroup.c:
```diff
@@ -0,0 +1,322 @@
+        TAILQ_INSERT_TAIL(&keyStorage->keyList, oldestKey, keyListEntry);
+        /*copy new key over the oldest key locatio*/
+        oldestKey->keyID = newKeyID;
+        UA_ByteString_copy(&newKey, &oldestKey->key);
```
Each security Group on SKS is configured with MaxFutureKeyCount and MaxPastKeyCount to define the maximum number of keys SKS can store with additional CurrentKey. So the oldest key is the maximum key we can store in the keystorage. So when we reach the maximum key size, we reuse the oldest key pointer for the new key and move it to the tail of the list.

Muddasir-Shakil replied:
Thanks for pointing out, I found a possible memleak. So after copying the new key data on the oldest key, I simply frees the pointer used to store the newkey data. This way we can avoid mem leak. In second case we anyhow use the newKey pointer in the keyStorage list.

Muddasir-Shakil commented:
hi @jpfr 
I updated the PR for thread safe adding of securityGroups, fix a memory leak, and passing securitygroup object. More importantly I fixed the commit messages according to the conventions
Looking forward to your feedback
